spec.html

```diff
@@ -7875,6 +7880,251 @@ <h1>Forward Progress</h1>
       <p>This, along with the liveness guarantee in the memory model, ensures that all ~SeqCst~ writes eventually become observable to all agents.</p>
     </emu-note>
   </emu-clause>
+
+  <emu-clause id="sec-weakref-processing-model">
+    <h1>Processing model of WeakRef and FinalizationRegistry objects</h1>
+    <emu-clause id="sec-weakref-invariants">
+      <h1>Objectives</h1>
+
+      <p>
+        This specification does not make any guarantees that any object will be garbage collected.
+        Objects which are unreachable from ECMAScript may be released after long periods
+        of time, or never at all. For this reason, this specification uses the term
+        "may" when describing behaviour triggered by garbage collection.
+      </p>
+
+      <p>
+        The semantics of WeakRef and FinalizationRegistry objects is based on two
+        operations which happen at particular points in time:
+      </p>
+
+      <ul>
+        <li>
+          When WeakRef.prototype.deref is called, the referent (if *undefined* is not returned)
+          is kept alive so that subsequent, synchronous accesses also
+          return the object. This list is reset when synchronous work is done using
+          the ClearKeptObjects abstract operation.
+        </li>
+
+        <li>
+          When an object which is registered with a FinalizationRegistry becomes
+          unreachable, a call of the FinalizationRegistry's cleanup callback may
+          eventually be made, after synchronous ECMAScript execution completes.
+          The FinalizationRegistry cleanup is performed with the
+          CleanupFFinalizationRegistry abstract operation.
+        </li>
+      </ul>
+
+      <p>
+        Neither of these actions (ClearKeptObjects or CleanupFinalizationRegistry)
+        may interrupt synchronous ECMAScript execution. Because embedding
+        environments may assemble longer, synchronous ECMAScript execution runs,
+        this specification defers the scheduling of ClearKeptObjects and
+        CleanupFinalizationRegistry to the embedding environment.
```

Comment:
```suggestion
        Neither of these actions (ClearKeptObjects or CleanupFinalizationRegistry) may interrupt synchronous ECMAScript execution. Because embedding environments may assemble longer, synchronous ECMAScript execution runs, this specification defers the scheduling of ClearKeptObjects and CleanupFinalizationRegistry to the embedding environment.
```